### PR TITLE
[#171706038] API for manual refresh idps metadata

### PR DESCRIPTION
**Pull request by BurnedMarshal**

### CODE REVIEW

File: api-metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+host: localhost
+basePath: /api/v1
+schemes:
+  - https
+security:
+  - Token: []
+paths:
+  "/idp-metadata-update":
+    x-swagger-router-controller: IdpMetadataController
+    get:
```
Comment (gunzip) [inline api-metadata.yaml]: can we have a post instead ?

File: api-metadata.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      operationId: refresh
```
Comment (gunzip) [inline api-metadata.yaml]:
```suggestion
      operationId: refreshIdpMetadata
```

File: api-metadata.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      summary: Update IDP Metadata
+      description: |-
+        Fetch IDP Metadata from remote url and update SPID Strategy
```
Comment (gunzip) [inline api-metadata.yaml]:
```suggestion
        Fetch IDP Metadata from remote registries then update the SPID Strategy
```

File: package.json [hunk 4]
```diff
@@ -48,7 +48,7 @@
   },
   "homepage": "https://github.com/pagopa/io-backend#readme",
   "dependencies": {
-    "@pagopa/io-spid-commons": "~2.9.1",
+    "@pagopa/io-spid-commons": "~3.0.0",
```
Comment (gunzip) [inline package.json]: why not ^3.0.0 ?

File: src/types/commons.ts [hunk 5]
```diff
@@ -12,3 +12,7 @@ export type SuccessResponse = t.TypeOf<typeof SuccessResponse>;
 
 export const STRINGS_RECORD = t.record(t.string, t.string);
 export type STRINGS_RECORD = t.TypeOf<typeof STRINGS_RECORD>;
+
+// Type from passport Authenticator used as response of session and authenticate methods
+// tslint:disable-next-line: no-any
+export type AuthenticateRet = any;
```
Comment (gunzip) [inline src/types/commons.ts]: why this alias ?

File: src/controllers/idpMetadataController.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  IResponseSuccessJson,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { SuccessResponse } from "../types/commons";
+
+export default class IdpMetadataController {
+  constructor(private readonly idpMetadataRefresher: () => Task<void>) {}
+
+  public async refresh(): Promise<IResponseSuccessJson<SuccessResponse>> {
+    await this.idpMetadataRefresher().run();
```
Comment (gunzip) [inline src/controllers/idpMetadataController.ts]: can you do this inside a setTimeout in order to free the socket ?

**Comment by gunzip** [comment]
this comment here: https://github.com/pagopa/io-backend/blob/master/src/config.ts#L297

must be changed to: "Token needed to receive API calls (notifications, metadata update) from io-functions-services"

File: api-metadata.yaml (hunk 1 above)
Comment (BurnedMarshal) [reply api-metadata.yaml]: Fixed in b6677a7

File: api-metadata.yaml (hunk 2 above)
Comment (BurnedMarshal) [reply api-metadata.yaml]: Fixed in b6677a7

File: api-metadata.yaml (hunk 3 above)
Comment (BurnedMarshal) [reply api-metadata.yaml]: Fixed in b6677a7

File: package.json (hunk 4 above)
Comment (BurnedMarshal) [reply package.json]: Fixed in b6677a7

File: src/types/commons.ts (hunk 5 above)
Comment (BurnedMarshal) [reply src/types/commons.ts]: `AuthenticateRet` is not exported into `@types/passport` and his value is `any`. We can use a pure `any` instead to create an alias. What do you suggest @gunzip ?

File: src/controllers/idpMetadataController.ts (hunk 6 above)
Comment (BurnedMarshal) [reply src/controllers/idpMetadataController.ts]: Async upgrade added here e2acd7c
Remove the `await` could be the same

File: src/types/commons.ts (hunk 5 above)
Comment (gunzip) [reply src/types/commons.ts]: that's ok to use a pure any[](url)

File: src/controllers/idpMetadataController.ts (hunk 6 above)
Comment (gunzip) [reply src/controllers/idpMetadataController.ts]: I'm not sure it's the same (danging promise vs timeout). It depends on how exception are handled; it depends on the context but usually unhandled exception in dangling promises could cause a program crash

Comment (gunzip) [reply src/controllers/idpMetadataController.ts]: probably it's ok to remove the await but yuo need to introduce a catch()

Comment (BurnedMarshal) [reply src/controllers/idpMetadataController.ts]: `idpMetadataRefresher` is a `Task` that returns a Promise that never catches or it should be. Must be caught anyway?

File: src/controllers/idpMetadataController.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { SuccessResponse } from "../types/commons";
+import { log } from "../utils/logger";
+
+export default class IdpMetadataController {
+  constructor(private readonly idpMetadataRefresher: () => Task<void>) {}
+
+  public async refresh(): Promise<IResponseSuccessJson<SuccessResponse>> {
+    // tslint:disable-next-line: no-floating-promises
```
Comment (gunzip) [inline src/controllers/idpMetadataController.ts]: Probably if we add a catch() we don't need to disable linting

File: src/controllers/idpMetadataController.ts (hunk 6 above)
Comment (gunzip) [reply src/controllers/idpMetadataController.ts]: yes it's always a good practice to add a catch()

File: src/controllers/idpMetadataController.ts (hunk 7 above)
Comment (BurnedMarshal) [reply src/controllers/idpMetadataController.ts]: Fixed in 8019d63

File: src/controllers/idpMetadataController.ts (hunk 6 above)
Comment (BurnedMarshal) [reply src/controllers/idpMetadataController.ts]: Fixed in 8019d63

File: src/types/commons.ts (hunk 5 above)
Comment (BurnedMarshal) [reply src/types/commons.ts]: Fixed in 8019d63

**Pull request merged**